### PR TITLE
chore: automate basic standards & document

### PR DESCRIPTION
**Pull request by portenez**
**Describe your changes**
Enable commit lint for commits

**Additional context**
commit lint makes sure all commit messages align to the [conventional-commits standard](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

File: .github/workflows/commitlint.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name: Lint Commit Messages
+on:
+  push:
+    branches: [main]
```
Comment (dakotalillie) [inline .github/workflows/commitlint.yml]: **Nitpick**

Do we need to lint when pushing to main? Or only on PR?

File: doc/architecture/decisions/0002-use-commitlint.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  > The Conventional Commits specification is a lightweight convention
+  > on top of commit messages. It provides an easy set of rules for creating
+  > an explicit commit history; which makes it easier to write automated tools on
+  > top of. This convention dovetails with SemVer, by describing the features,
+  > fixes, and breaking changes made in commit messages.
+
+- [Commitlint](https://github.com/conventional-changelog/commitlint) is
+  a tool for enforcing commit messages
+  
+  *from the `commitlin` github repo*:
```
Comment (dakotalillie) [inline doc/architecture/decisions/0002-use-commitlint.md]:
```suggestion
  *from the `commitlint` github repo*:
```

File: doc/architecture/decisions/0002-use-commitlint.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  > commitlint checks if your commit messages meet the conventional commit format.
+
+## Decision
+
+- We will use `conventional-commits` in this project.
+- We will enforce `conventional-commits` through `commitlint`
+- We will enforce `conventional-commits` in PRs through a `Github action`.
```
Comment (dakotalillie) [inline doc/architecture/decisions/0002-use-commitlint.md]: **Nitpick**

```suggestion
- We will enforce `conventional-commits` through the `commitlint` GitHub action.
```

File: doc/architecture/decisions/0002-use-commitlint.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+- Using an automated tools for releases. e.g.: [semantic-release](https://github.com/semantic-release/semantic-release)
+- Keeping a uniform format for commit messages
+- Reviewing PRs, because reviewers will not need to worry about also reviewing
+  commit message format.
+
+### What becomes more difficult
+
+- Pushing quick (wip) commits to a pull request will be more difficult
+  as contributors will get need to know to use the right format.
```
Comment (dakotalillie) [inline doc/architecture/decisions/0002-use-commitlint.md]:
```suggestion
  as contributors will need to learn to use the right format.
```

File: doc/architecture/decisions/0002-use-commitlint.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+### Risks introduced and mitigation
+
+| Risk                                              | Mitigation                                                                               |
+| ------------------------------------------------- | ---------------------------------------------------------------------------------------- |
+| The automated PR check can frustrate contributors | Clearly outline in the README that constributors should use conventional commit messages |
```
Comment (dakotalillie) [inline doc/architecture/decisions/0002-use-commitlint.md]:
```suggestion
| The automated PR check can frustrate contributors | Clearly outline in the README that contributors should use conventional commit messages |
```

**Review by dakotalillie — COMMENTED** [review]
Perhaps it's out of scope, but — would we also want to also use the conventional commit format for PR titles, limit to squash commits, and set up a branch protection rule on `main` which prevents force pushes and requires a linear history (no merge commits)?

File: .github/workflows/commitlint.yml (hunk 1 above)
Comment (portenez) [reply .github/workflows/commitlint.yml]: This will lint both on main and PR.  We need to lint on PRs. I'm not sure about main. What do you think?

**Comment by portenez** [comment]
the DCO check.. I'm going to pull, then squash with the missing check. 

Also, I'm thinking, perhaps we should also add an automated spell checker. I'll see if I can add that now.

**Comment by portenez** [comment]
Let's look for a spellchecker GitHub action later.  I've corrected the DCO.

File: .github/workflows/commitlint.yml (hunk 1 above)
Comment (dakotalillie) [reply .github/workflows/commitlint.yml]: My vote is only on PRs, not on main.

Comment (portenez) [reply .github/workflows/commitlint.yml]: fine with me. I'll update

**Comment by portenez** [comment]
> and set up a branch protection rule on main

this is already set up. 

> would we also want to also use the conventional commit format for PR titles

yes, is there a way to enforce? or should we also say that in the `README.md`?

**Comment by dakotalillie** [comment]
> yes, is there a way to enforce? or should we also say that in the `README.md`?

I believe there is a way to enforce it in PR titles — we use something for that on BB-UI. Can look into it more later, hackathon is still ongoing.

**Comment by dakotalillie** [comment]
Another major nitpick, but it's my impression the directory is usually called `docs`, not `doc`.

**Review by dakotalillie — APPROVED** [review]
🎉

**Pull request merged**